shell/common/asar/archive_mac.mm

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/common/asar/archive.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <Foundation/Foundation.h>
+
+#include <string>
+
+#include "base/logging.h"
+#include "base/mac/foundation_util.h"
+#include "base/strings/sys_string_conversions.h"
+#include "electron/fuses.h"
+
+namespace asar {
+
+void Archive::MaybeValidateArchiveSignature() {
+// This method uses a private API SecCodeValidateFileResource
+// which should not be used in the MAS build
+#ifndef MAS_BUILD
+  // If the fuse is not enabled, don't waste any time here.
+  if (!electron::fuses::IsEmbeddedAsarSignatureValidationEnabled())
+    return;
+
+  NSURL* bundle_url =
+      [NSURL URLWithString:[[NSBundle mainBundle] executablePath]];
+  // /Stuff/Electron.app/Contents/MacOS/Electron is the bundle url.
+  NSURL* contents_url = [[bundle_url URLByDeletingLastPathComponent]
+      URLByDeletingLastPathComponent];
+  const std::string abs_path = path_.value();
+
+  // If the path to the app.asar file is not contained by the bundle we can't
+  // validate the signature.
+  if (abs_path.rfind(base::SysNSStringToUTF8([contents_url absoluteString]),
+                     0) != 0)
+    return;
+
+  SecStaticCodeRef ref = NULL;
+
+  OSStatus status;
+  status = SecStaticCodeCreateWithPath((CFURLRef)bundle_url, kSecCSDefaultFlags,
+                                       &ref);
+
+  if (ref == NULL) {
+    LOG(FATAL) << "Failed to create static code reference for " << bundle_url;
+    return;
+  }
+
+  if (status != noErr) {
+    CFRelease(ref);
+    LOG(FATAL) << "Failed to create static code reference for " << bundle_url;
+    return;
+  }
+
+  NSString* archive_path = base::mac::FilePathToNSString(path_);
+  NSString* relative_path =
+      [archive_path substringFromIndex:[[contents_url absoluteString] length]];
+  NSData* data = [NSData dataWithContentsOfFile:archive_path];
+  status = SecCodeValidateFileResource(ref, (CFStringRef)relative_path,
+                                       (CFDataRef)data, kSecCSDefaultFlags);
```

Comment:
Chrome seems to do some slightly different stuff: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/safe_browsing/signature_evaluator_mac.mm;l=216;drc=8bee0395f4d84e89e7dfa0821880f48643426008

thoughts?

Reply:
They appear to validate _the entire bundle_ 🤔  `SecStaticCodeCheckValidityWithErrors` validates the entire thing not just a single executable, we specifically want to avoid validating 300MB worth of application as the OS is already doing that via hardened runtime, we only need to validate the small part of the app that is the ASAR file.

That code looks like it was written before hardened runtime became a thing, nowadays if that check failed in Chrome the app would just hard crash courtesy of macOS.